src/test/basics/StringUtilities_test.cpp

```diff
@@ -63,23 +63,133 @@ class StringUtilities_test : public beast::unit_test::suite
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://:pass@domain:123/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "");
+            BEAST_EXPECT(pUrl.password == "pass");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(*pUrl.port == 123);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://domain:123/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "");
+            BEAST_EXPECT(pUrl.password == "");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(*pUrl.port == 123);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://user:pass@domain/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "user");
+            BEAST_EXPECT(pUrl.password == "pass");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(! pUrl.port);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://user@domain/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "user");
+            BEAST_EXPECT(pUrl.password == "");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(! pUrl.port);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://:pass@domain/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "");
+            BEAST_EXPECT(pUrl.password == "pass");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(! pUrl.port);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme://domain/abc:321"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "");
+            BEAST_EXPECT(pUrl.password == "");
+            BEAST_EXPECT(pUrl.domain == "domain");
+            BEAST_EXPECT(! pUrl.port);
+            BEAST_EXPECT(pUrl.path == "/abc:321");
+        }
+        {
+            parsedURL pUrl;
+            BEAST_EXPECT(parseUrl (pUrl, "scheme:///path/to/file"));
+            BEAST_EXPECT(pUrl.scheme == "scheme");
+            BEAST_EXPECT(pUrl.username == "");
+            BEAST_EXPECT(pUrl.password == "");
+            BEAST_EXPECT(pUrl.domain == "");
+            BEAST_EXPECT(! pUrl.port);
+            BEAST_EXPECT(pUrl.path == "/path/to/file");
+        }
```

Comment:
Another couple of suggestions for test cases:
* `scheme://user:pass@domain/path/with/an@sign`
* `scheme://domain/path/with/an@sign`
* `scheme://`
* `scheme:///`

And we need some negative test cases, too:
* `(empty string)`
* `nonsense`
* `://`
* `:///`
* `scheme://:999/`

Reply:
That last test, `scheme://:999/`, is not a negative case. [RFC 3986 allows an empty host with a non-empty port.](https://tools.ietf.org/html/rfc3986#appendix-A) (You have to dig into the rule for `reg-name` to see that `host` can be empty, and the rule for `authority` shows that the presence of `port` is independent of `host`.)

Reply:
TIL!

And I see that you made it a positive case, which is great.